### PR TITLE
Games: Fix launching disk images and zip files

### LINKED ISSUE #15841: Kodi Ignores videoextensions/remove advancedsetting for iso
### Describe the bug
Here is a clear and concise description of what the problem is:

Attempting to play an iso file using retroplayer.  I have the following advancedsettings.xml file:
```
<advancedsettings>
<videoextensions>
<remove>.cue|.bin|.img|.iso|.zip</remove>
</videoextensions>
<musicextensions>
<remove>.cue|.bin|.img|.iso|.zip</remove>
</musicextensions>
<pictureextensions>
<remove>.cue|.bin|.img|.iso|.zip</remove>
</pictureextensions>
</advancedsettings>
```

But when selecting an ISO file, Kodi still attempts to play it as a video file.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Kodi removes the iso extension as a valid videofile extension with the above adancedsettings.xml

## Actual Behavior

With the advancedsettings.xml file present, Kodi still tries to play the file as a video file.

### To Reproduce

1.  Use the advancedsettings.xml pasted above
2.  In the v18 Kodi games section, select an iso game file
3. Note that there is a flash as Kodi tries to open the file as a video, then returns since it's not a valid video file

### Debuglog

The debuglog can be found here:
```
15:06:39.000 T:4747515328  NOTICE: Loaded settings file from special://profile/advancedsettings.xml
15:06:39.000 T:4747515328  NOTICE: Contents of special://profile/advancedsettings.xml are...
                                            <advancedsettings>
                                              <videoextensions>
                                                <remove>.cue|.bin|.img|.iso|.zip</remove>
                                              </videoextensions>
                                              <musicextensions>
                                                <remove>.cue|.bin|.img|.iso|.zip</remove>
                                              </musicextensions>
                                              <pictureextensions>
                                                <remove>.cue|.bin|.img|.iso|.zip</remove>
                                              </pictureextensions>
                                            </advancedsettings>

...

15:06:52.511 T:4747515328   DEBUG: HandleKey: return (0xf00d) pressed, action is Select
15:06:52.513 T:4747515328   DEBUG: CPlayerCoreFactory::GetPlayers(/Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/ReturnFireeurope.iso)
15:06:52.513 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: system rules
15:06:52.513 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: system rules
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: streams
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: dvd
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: dvdimage
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: dvdimage
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: adding player: VideoPlayer for rule: dvdimage
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: nsv
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: radio
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: un-named
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: un-named
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: retroplayer_iso
15:06:52.514 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: retroplayer_iso
15:06:52.515 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: retroplayer_m3u
15:06:52.515 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: retroplayer_cue
15:06:52.515 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: retroplayer_bin
15:06:52.515 T:4747515328   DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: retroplayer_img
15:06:52.515 T:4747515328   DEBUG: CPlayerCoreFactory::GetPlayers: matched 1 rules with players
15:06:52.515 T:4747515328   DEBUG: CPlayerCoreFactory::GetPlayers: adding retroplayer
15:06:52.515 T:4747515328   DEBUG: CPlayerCoreFactory::GetPlayers: added 2 players
15:06:52.517 T:4747515328   DEBUG: Radio UECP (RDS) Processor - new CDVDRadioRDSData
15:06:52.517 T:4747515328  NOTICE: VideoPlayer::OpenFile: /Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/ReturnFireeurope.iso
15:06:52.517 T:4747515328   DEBUG: OnPlayBackStarted: CApplication::OnPlayBackStarted
15:06:52.517 T:123145385136128   DEBUG: Thread VideoPlayer start, auto delete: false
15:06:52.518 T:123145385136128  NOTICE: Creating InputStream
15:06:52.519 T:123145385136128   DEBUG: SECTION:LoadDLL(special://xbmcbin/system/players/VideoPlayer/libdvdnav-x86-osx.so)
15:06:52.519 T:123145385136128   DEBUG: Loading: /Applications/Kodi_v18.app/Contents/Resources/Kodi/system/players/VideoPlayer/libdvdnav-x86-osx.so
15:06:52.521 T:123145385136128    INFO:   msg: libdvdnav: Using dvdnav version 6.0.0
15:06:52.522 T:4747515328   DEBUG: CPlayerGUIInfo::InitCurrentItem(/Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/ReturnFireeurope.iso)
15:06:52.523 T:123145407041536   DEBUG: Loading settings for /Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/ReturnFireeurope.iso
15:06:52.523 T:4747515328   DEBUG: CGamesGUIInfo::InitCurrentItem(/Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/ReturnFireeurope.iso)
15:06:52.523 T:123145387819008   DEBUG: CAnnouncementManager - Announcement: OnChanged from xbmc
15:06:52.523 T:123145387819008   DEBUG: GOT ANNOUNCEMENT, type: 1024, from xbmc, message OnChanged
15:06:52.523 T:123145387819008   DEBUG: CAnnouncementManager - Announcement: OnPlay from xbmc
15:06:52.523 T:123145387819008   DEBUG: GOT ANNOUNCEMENT, type: 1, from xbmc, message OnPlay
15:06:52.525 T:123145385136128    INFO:   msg: libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
15:06:52.525 T:123145385136128    INFO:   msg: libdvdread:DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.BUP failed
15:06:52.525 T:123145385136128    INFO:   msg: libdvdread: Can't open file VIDEO_TS.IFO.
15:06:52.525 T:123145385136128    INFO:   msg: libdvdnav: vm: failed to read VIDEO_TS.IFO
15:06:52.525 T:123145385136128   ERROR: Error opening image file or Error on dvdnav_open_stream
```

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

I will note that I also attempted to make a player rule to force it to use retroplayer, and that also does not seem to work when the rule action is 'append':
```
<playercorefactory>
<rules action="append">
<rule name="retroplayer_iso" filetypes="iso" player="retroplayer"/>
<rule name="retroplayer_m3u" filetypes="m3u" player="retroplayer"/>
<rule name="retroplayer_cue" filetypes="cue" player="retroplayer"/>
<rule name="retroplayer_bin" filetypes="bin" player="retroplayer"/>
<rule name="retroplayer_img" filetypes="img" player="retroplayer"/>
</rules>
</playercorefactory>
```
If I change it to rule action 'overwrite' and then it seems to override the videoplayer completely and does use retroplayer.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] Linux
 - [X] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version:  V18.1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by garbear**
## Description
This PR fixes two errors when launching games. Closes #15841.

### First fix

The first fix is for disk images. In `system/playercorefactory.xml` is the line:

```xml
<rule name="dvdimage" dvdimage="true" player="VideoPlayer" />
```

Setting `dvdimage="true"` calls into FileItem [here](https://github.com/xbmc/xbmc/blob/f4d6b4022/xbmc/FileItem.cpp#L1087) to see if the extension is a disk image: 

```c++
bool CFileItem::IsDiscImage() const
{
  return URIUtils::HasExtension(m_strPath, ".img|.iso|.nrg|.udf");
}
```

We want to avoid using VideoPlayer for fileitems with a game info tag, so I've modified the xml line to be:

```xml
<rule name="dvdimage" dvdimage="true" game="false" player="VideoPlayer" />
```

This skips the `IsDiscImage()` comparison for FileItems with a game info tag.

### Second fix

The second fix is for zip and 7z games not launching. They aren't matched to a player and default to VideoPlayer. When launched, VideoPlayer shows a buffering window then immediately exits to the previous window.

I fixed this by giving ListItems a game info tag in MyGames. Now, when no rules match, RetroPlayer is used to open the file.

## Motivation and Context

See #15841

## How Has This Been Tested?

Test builds: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.5-20200127

### Launching a .zip file

Before:

```
DEBUG: CPlayerCoreFactory::GetPlayers(/Users/xxx/Library/Application Support/Kodi/userdata/addon_data/plugin.program.iagl/temp_iagl/88games.7z)
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: system rules
DEBUG: CPlayerSelectionRule::GetPlayers: matches rule: system rules
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: streams
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: dvd
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: nsv
DEBUG: CPlayerSelectionRule::GetPlayers: considering rule: radio
DEBUG: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
DEBUG: CPlayerCoreFactory::GetPlayers: adding videodefaultplayer (VideoPlayer)
DEBUG: CPlayerCoreFactory::GetPlayers: for video=1, audio=0
DEBUG: CPlayerCoreFactory::GetPlayers: for video=1, audio=1
DEBUG: CPlayerCoreFactory::GetPlayers: added 1 players
```

After:

```
Debug Print: CPlayerCoreFactory::GetPlayers(C:\Users\xxx\Desktop\temp_iagl\88games.zip)
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: system rules
Debug Print: CPlayerSelectionRule::GetPlayers: matches rule: system rules
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: mms/udp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: lastfm/shout
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: rtmp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: rtsp
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: streams
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: dvd
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: dvdimage
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: sdp/asf
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: nsv
Debug Print: CPlayerSelectionRule::GetPlayers: considering rule: radio
Debug Print: CPlayerCoreFactory::GetPlayers: matched 0 rules with players
Debug Print: CPlayerCoreFactory::GetPlayers: adding retroplayer
Debug Print: CPlayerCoreFactory::GetPlayers: added 1 players
...
Debug Print: Select game client dialog: Found 9 candidates
Debug Print: Adding game.libretro.cap32 as a candidate
Debug Print: Adding game.libretro.fbalpha2012 as a candidate
Debug Print: Adding game.libretro.mame2000 as a candidate
Debug Print: Adding game.libretro.mame2010 as a candidate
Debug Print: Adding game.libretro.hatari as a candidate
Debug Print: Adding game.libretro.uae as a candidate
Debug Print: Adding game.libretro.vice as a candidate
Debug Print: Adding game.libretro.reicast as a candidate
Debug Print: Adding game.libretro.yabause as a candidate
Debug Print: Select game client dialog: Found 1 installable clients
Debug Print: Adding game.libretro.fbalpha as an installable client
Debug Print: Select game client dialog: Loading savestate metadata
Debug Print: Loading savestate from C:\Users\xxx\Desktop\temp_iagl\88games.sav
Debug Print: Failed to open savestate file C:\Users\xxx\Desktop\temp_iagl\88games.sa
```

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by da-anda** [comment]
while I don't mind having an additional game condition possibility here, I think the logic behind `dvdimage=true` needs to be adjusted instead. It clearly says **DVD**image and not **DISC**image and should certainly return `false` for anything but DVDs (which also excludes blu-rays IMO). So the additional game condition should not be needed here at all.

File: xbmc/games/windows/GUIWindowGames.cpp [hunk 1]
```diff
@@ -265,6 +265,14 @@ bool CGUIWindowGames::GetDirectory(const std::string &strDirectory, CFileItemLis
   if (!content.empty())
     items.SetContent(content);
 
+  // Ensure files have a game info tag so that they're recognized as games
+  for (auto it = items.begin(); it != items.end(); ++it)
```
Comment (Rechi) [inline xbmc/games/windows/GUIWindowGames.cpp]: use a range based for loop

Comment (garbear) [reply xbmc/games/windows/GUIWindowGames.cpp]: `items` is a `CFileItemList`, I tried to figure out how to range-based loop over it but couldn't. Do you know anywhere this is done?

**Comment by garbear** [comment]
> while I don't mind having an additional game condition possibility here, I think the logic behind dvdimage=true needs to be adjusted instead. It clearly says DVDimage and not DISCimage and should certainly return false for anything but DVDs (which also excludes blu-rays IMO). So the additional game condition should not be needed here at all.

Hard to change the logic without breaking anything. What if we change the name from `dvdimage` to `diskimage`?

File: xbmc/games/windows/GUIWindowGames.cpp (hunk 1 above)
Comment (Rechi) [reply xbmc/games/windows/GUIWindowGames.cpp]: I don't know if it is already done anywhere else, but if you look at
https://github.com/xbmc/xbmc/blob/f4d6b4022fc3122471f0d39aeb58abafc9412a10/xbmc/FileItem.h#L761-L766
and
https://github.com/xbmc/xbmc/blob/f4d6b4022fc3122471f0d39aeb58abafc9412a10/xbmc/FileItem.h#L598
you'll come to the following result.
```cpp
  for (const CFileItemPtr& item : items)
  {
    if (!item->m_bIsFolder)
      item->GetGameInfoTag();
  }

```

**Comment by jjd-uk** [comment]
I think there's a good chance it was called DVDimage simply because the player was called DVDPlayer up until fernet changed the name.

**Comment by da-anda** [comment]
renaming it to `diskimage` would also be fine (while keeping the old name for backwards compatibility for at least 1 release). Would at least be way less confusing.

File: xbmc/games/windows/GUIWindowGames.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/games/windows/GUIWindowGames.cpp]: updated

**Comment by garbear** [comment]
@da-anda I renamed to `discimage` and added a deprecation warning

**Comment by lrusak** [comment]
@garbear you have a build failure :)

```
08:16:39 /home/jenkins/workspace/LINUX-64/xbmc/games/windows/GUIWindowGames.cpp: In member function ‘virtual bool KODI::GAME::CGUIWindowGames::GetDirectory(const string&, CFileItemList&)’:
08:16:39 /home/jenkins/workspace/LINUX-64/xbmc/games/windows/GUIWindowGames.cpp:269:29: error: binding reference of type ‘CFileItemPtr& {aka std::shared_ptr<CFileItem>&}’ to ‘const std::shared_ptr<CFileItem>’ discards qualifiers
08:16:39    for (CFileItemPtr& item : items)
08:16:39                              ^~~~~
```

**Comment by garbear** [comment]
@Rechi I modified the comment:

```
// Ensure a game info tag is created so that files are recognized as games
```

I need points to mutable CFileItems. It looks like the iters you linked are all const. Should I add non-const iterators, or is there a way to iterate over mutable items?

**Comment by Rechi** [comment]
There seems to be something wrong with the (c)begin and (c)end functions of the CFileItemList class. The following should be the fix.
```diff
--- a/xbmc/FileItem.h
+++ b/xbmc/FileItem.h
@@ -760,8 +760,8 @@
 
-  VECFILEITEMS::const_iterator begin() { return m_items.cbegin(); }
-  VECFILEITEMS::const_iterator end() { return m_items.cend(); }
+  VECFILEITEMS::iterator begin() { return m_items.begin(); }
+  VECFILEITEMS::iterator end() { return m_items.end(); }
   VECFILEITEMS::const_iterator begin() const { return m_items.begin(); }
   VECFILEITEMS::const_iterator end() const { return m_items.end(); }
-  VECFILEITEMS::const_iterator cbegin() const { return m_items.begin(); }
-  VECFILEITEMS::const_iterator cend() const { return m_items.end(); }
+  VECFILEITEMS::const_iterator cbegin() const { return m_items.cbegin(); }
+  VECFILEITEMS::const_iterator cend() const { return m_items.cend(); }
 private:
```

Your actual issue is that you use `for (CFileItemPtr& item : items)` instead of `for (const CFileItemPtr& item : items)`, like in my code suggestion.
`CFileItemPtr` == `std::shared_ptr<CFileItem>`, so it's `for (const std::shared_ptr<CFileItem>& item : items)`, which means that only the shared pointer, but not the `CFileItem` itself, is const and the correct (non const) variant of `GetGameInfoTag` gets called.

**Comment by garbear** [comment]
That looked a little strange to me, I'll include the fix in this PR. And you make a good point, I'll add const

**Comment by garbear** [comment]
Updated

**Comment by garbear** [comment]
jenkins build this please

**Pull request merged**